sources/xfscli/src/bin/fsck_xfs.rs

```diff
@@ -0,0 +1,169 @@
+    let force = args.force;
+    let repair = args.repair;
+
+    if let Some(target) = &args.target {
+        if !Path::new(target).exists() {
+            println!("{} does not exist", target);
+            exit(8);
+        }
+
+        if force {
+            // Preform xfs_repair
+            println!("Performing XFS repair..");
+            let mut repair_exit_code = perform_xfs_repair(vec!["-e", target]);
+
+            // Clear the log, mount and unmount the XFS file system
+            // before xfs_repair
+            if repair_exit_code == 2 && repair {
+                println!("Replaying log for {}", target);
+                let mkdir_result = fs::create_dir_all("/tmp/repair_mnt");
+                if mkdir_result.is_err() {
+                    exit(1);
+                }
+
+                let params =
+                    fs::read_to_string("/proc/cmdline").expect("Unable to read /proc/cmdline file");
+                let mut root = String::new();
+                let mut root_flags = "-o ".to_string();
+                for param in params.split(' ') {
+                    if param.starts_with("root=") {
+                        root = param.strip_prefix("root=").unwrap().to_string();
+                    }
+                    if param.starts_with("rootflags=") {
+                        root_flags = root_flags
+                            .clone()
+                            .add(param.strip_prefix("rootflags=").unwrap());
+                    }
+                }
+                let path = Path::new(&root);
+                let metadata = path.metadata().expect("Could not fetch metadata");
+                let file_type = metadata.file_type();
+
+                if !file_type.is_block_device() {
+                    let blkid_result = Command::new(BLKID)
+                        .args(["-t", &root, "-o", target])
+                        .output()
+                        .expect("Couldn't run blkid");
+                    root = String::from_utf8(blkid_result.stdout).unwrap();
+                }
+
+                let basename_dev = Path::new(target).file_name().unwrap();
+
+                let basename_root = Path::new(&root).file_name().unwrap();
+
+                let mut mount_args = vec![target.as_str(), "/tmp/repair_mnt"];
+                if basename_dev == basename_root {
+                    mount_args.push(root_flags.as_str());
+                }
+
+                let mount_result = Command::new(MOUNT)
+                    .args(mount_args)
+                    .output()
+                    .expect("Couldn't run mount");
+                if !mount_result.status.success() {
+                    exit(1);
+                }
+
+                Command::new(UMOUNT)
+                    .args(["/tmp/repair_mnt"])
+                    .output()
+                    .expect("Couldn't run unmount");
+
+                println!("Performing XFS repair again after cleaning the log..");
```

Comment:
Is the log cleared basically by mounting/unmounting?

Reply:
As per my understanding the [script](https://git.kernel.org/pub/scm/fs/xfs/xfsprogs-dev.git/tree/fsck/xfs_fsck.sh) is replaying the logs by mounting and unmounting the device when repair is unsuccessful . @markusboehme / @foersleo  Will it be possible for you to share your thoughts on this?

Reply:
As per my understanding the [script](https://git.kernel.org/pub/scm/fs/xfs/xfsprogs-dev.git/tree/fsck/xfs_fsck.sh) is replaying the logs by mounting and unmounting the device when repair is unsuccessful . @markusboehme / @foersleo  Will it be possible for you to share your thoughts on this?

Reply:
Yes, a mount/unmount cycle is what's needed here. XFS is a journaling file system, i.e. it records intended changes to the file system before actually performing them. Should the IO be interrupted (power loss, machine crashes, disk is unresponsive, etc.) the file system can still be brought into a consistent state by performing the intended actions on the next mount ("replaying/clearing the journal"). The actual work here is performed by the mount; the unmount just needs to happen eventually so the device can be touched directly without interfering with normal accesses.